### ISSUE TITLE
Checkout : Shipping before payment

### ISSUE THREAD

**Issue by Zmove** [OFFLINE-GmbH/oc-mall-plugin#303]
Hello,

I tested the demo, both frontend and backend, and it's really well done, congratulation to the author.

However, in checkout, is it possible to put shipping method choice step before payment ? It sounds more logical to me in that order.

Maybe as an option in module settings to not annoy people that would prefer payment before shipping.

Thank you for your great work

Best regards,

Alex

**Comment by tobias-kuendig** [comment]
We might change this or make it configurable but for the time being this is a very low priority feature.

If you really must change the order of the steps you can extend the relevant components and override the `redirect` methods on all of them.

**Comment by damsfx** [comment]
I'm surprised!   

In many cases, the price to paid includes shipping taxes.  
These taxes vary depending on the shipping method chosen (postal services vs. express services) and the destination (host country vs. foreign country).  

How can you charge for a delivery service before you have chosen it?  
Without necessarily thinking of international trade, shipping costs can also vary for different regions of the same country: for example with France and its overseas departments.

**Comment by tobias-kuendig** [comment]
> The entry of bank information and the validation of the payment must take place after the choice of transport and a summary of the basket.

With the current setup, the payment gateway is not contacted during the selection of the payment method. You choose the payment method, then the shipping method and then you'll see your cart again with the effective price that includes all country specific taxes.

Only once you create the order, the payment gateway will be contacted and the payment is done. In case the payment fails, the order remains in an unpaid state and the customer can retry to pay again.

**Comment by damsfx** [comment]
I understood it that way.  

But when you look at e-commerce sites, very often the entry of information for payment does not take place until you have chosen the shipping costs.  
It's silly but from experience, customers like to have a summary of their order and a nice "pay now" button.  

Like them, I find it destabilizing...   
Imagine... I want to have an estimate of the shipping taxes before ordering.  
In this case, I must enter my payment information before I have any information on shipping costs.    Not very encouraging.   ¯\\\_(ツ)\_/¯

**Comment by tobias-kuendig** [comment]
Well, I guess in the end it really comes down to a matter of preference. It is, however, rather trivial to change the order of the checkout steps by extending the default components and override the rewrites. 

A future release will include a "single page checkout" (#6) component that has all of the checkout steps on a single page.

**Comment by Zmove** [comment]
I think an option to put shipping step before payment in multi-step checkout would be useful, even with single page checkout component. Not a priority, but it can handle some needs.

**Comment by Hessel91** [comment]
you can always use a workaround by hiding the shipping data on the cart page and show it on the checkout page. Just add viewbag to the cart page. For example:
`[viewBag]
page = "cart"`

and then overwrite cart/cart.htm in your partial and modify:
```
{% if viewBag.page == 'cart' %}    
    <!--nothing to see here-->
  {% else %}   
    {% partial __SELF__ ~ '::table/shipping'   shipping=cart.totals.shippingTotal %}
{% endif %}
```

**Comment by lieszkol** [comment]
Added a PR to try and address this, but it needs to be reviewed. I hope it helps resolve this issue (which I ran into myself, my client asked for this exact change).

**Comment by tobias-kuendig** [comment]
Fixed via 700bb771f92df29ec3ba9476d75626a6f829b5a7, credits to @lieszkol
